c/tskit/tables.c

```diff
@@ -13003,3 +13003,164 @@ tsk_squash_edges(tsk_edge_t *edges, tsk_size_t num_edges, tsk_size_t *num_output
 out:
     return ret;
 }
+
+/* ======================================================== *
+ * Tree diff iterator.
+ * ======================================================== */
+
+int TSK_WARN_UNUSED
+tsk_diff_iter_init_tables(
+    tsk_diff_iter_t *self, const tsk_table_collection_t *tables, tsk_flags_t options)
+{
+    int ret = 0;
+
+    tsk_bug_assert(tables != NULL);
+    tsk_memset(self, 0, sizeof(tsk_diff_iter_t));
+    self->num_nodes = tables->nodes.num_rows;
+    self->num_edges = tables->edges.num_rows;
+    self->tables = tables;
+    self->insertion_index = 0;
+    self->removal_index = 0;
+    self->tree_left = 0;
+    self->tree_index = -1;
+    self->last_index
```

Comment:
This PR gives us two ways to initialize this iterator:

* from a tree sequence 
* from a table collection

Prior to this PR, `self->last_index` got pulled from the tree sequence's `num_trees` field.  But, since a table collection doesn't know how many trees it has, this PR always calls the table validator to get that value. The table validation is not necessary when the input is a table collection owned/referred to by a tree sequence -- those tables have already been validated.

Reply:
Right, gotcha. that is annoying

Reply:
So it seems we have a couple of options: use flags.  Currently, the flag is assumed to be zero, so we'd perhaps want a non-zero to say "count trees" and require that for tables.  However, that flag setup is actually the opposite of what is needed by default behavior and is impossible (?) to verify at run time (we cannot know where the table collection pointer comes from).  Need to ponder it a bit more.

Reply:
Another option is to somehow use an internal API to always "do the right thing".  That's where I need to think harder.

Reply:
It's a bit fiddly, but I guess we could change the signature of the init function to take ``num_trees`` as a parameter, which you pass ``-1`` if you don't know it. It's an internal only API not used that much, so not too terrible?

Reply:
That was one of my first thoughts.  I agree that it isn't too bad.  My concern is that it is public via C, even though it is undocumented.  The only way to tell the user that the value needs to be correct is via documentaton.

Reply:
It's an internal API and I don't think we should document it, so :shrug:

We can coordinate on not breaking tskit-rust I guess, maybe just with some comments to remind us (rather than actual docs)

Reply:
We have to have the freedom to change undocumented things, otherwise you go mad trying to write anything new

Reply:
Yeah, that works for me.  tskit-rust will just figure itself out -- I'll manage the constructor functions to do the right thing and any breakage will only happen internally, which is no big deal.